Added several new fields to the `ReadGroup` index. DSDEGP-1340

kshakir opened the pull request:
### Description

Interpreted the ticket as the _union_ with the existing `ReadGroup` index fields, not a replacement.
Open to other preferences / suggestions for orderings the index's fields.
DSDEGP-1340

Bonus:
- Fixed capitalization typo in `ElasticsearchIndexDefiner`.
- Fixed field type for `ElasticsearchIndexDefiners.All`.
- Index fields now also allow the nullable Java boxed types.
- Alphabetized the index fields.

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: Ensure that you have added or updated tests
- [x] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Move the JIRA issue to QA once this checklist is completed

tbl3rd commented on src/main/scala/org/broadinstitute/clio/model/indexes.scala:
```diff
@@ -1,18 +1,44 @@
 
 case class ReadGroup
 (
+  initiative: String,
```
What order are these in?

kshakir replied:
A combination of looking at the field order in the Picard dev database, and then guessing. Very open to a better order, but I don't think Elasticsearch nor Scala care.

tbl3rd replied:
Whenever I see something not in some obvious order (lexicographic, numeric, whatevz),
I wonder what I'm missing.

kshakir replied:
Fields are now alphabetized.

tbl3rd commented on src/main/scala/org/broadinstitute/clio/dataaccess/HttpElasticsearchDAO.scala:
```diff
@@ -1,5 +1,6 @@
 package org.broadinstitute.clio.dataaccess
 
+import java.lang.{Boolean => JBoolean, Double => JDouble, Float => JFloat, Integer => JInteger, Long => JLong}
 import java.time.OffsetDateTime
```
Weird. Where was OffsetDateTime used before?

kshakir replied:
It was available in the mapper, but wasn't being passed. I was just thinking ahead, as I didn't want future devs trying to leak in [Joda-Time](http://www.joda.org/joda-time/), or even java.util.Date.

tbl3rd replied:
Just surprised that scalac compiled it that way.

tbl3rd commented on src/main/scala/org/broadinstitute/clio/dataaccess/HttpElasticsearchDAO.scala:
```diff
@@ -130,12 +131,17 @@ object HttpElasticsearchDAO extends LazyLogging {
   private def fieldDefinition(elasticsearchField: ElasticsearchField): FieldDefinition = {
```
This is some type-safe glue code between Clio and the Elastic4s DSL?

kshakir replied:
Exactly! The `ElasticsearchField` was at one point an elastic4s `FieldDefinition`, but Phil pointed out that the Elasticsearch specific code should be hidden away. Now, `ElasticsearchField` might as well be called `TableColumn`-- there's nothing particularly ES specific about the data stored in each instance.

tbl3rd commented on src/main/scala/org/broadinstitute/clio/dataaccess/elasticsearch/ElasticsearchIndexDefiner.scala:
```diff
@@ -2,6 +2,6 @@ package org.broadinstitute.clio.dataaccess.elasticsearch
 import org.broadinstitute.clio.model.ElasticsearchIndex
 
-trait ElasticSearchIndexDefiner[A] {
+trait ElasticsearchIndexDefiner[A] {
```
Nice.

tbl3rd reviewed (approved):
LGTM, but lessons sought ...

kshakir committed:
Added several new fields to the `ReadGroup` index.
Fixed capitalization typo in `ElasticsearchIndexDefiner`.
Fixed field type for `ElasticsearchIndexDefiners.All`.
Index fields now also allow the nullable Java boxed types.
Alphabetized the index fields.
